api/keto/acl/v1/write_service.proto

```diff
@@ -0,0 +1,76 @@
+syntax = "proto3";
+
+package keto.acl.v1;
+
+import "keto/acl/v1/acl.proto";
+import "google/api/client.proto";
+
+option go_package = "github.com/ory/keto/api/keto/acl/v1;acl";
+option csharp_namespace = "Ory.Keto.Acl.V1";
+option java_multiple_files = true;
+option java_outer_classname = "WatchServiceProto";
+option java_package = "sh.ory.keto.acl.v1";
+option php_namespace = "Ory\\Keto\\Acl\\V1";
+
+// The write service to create and delete Access Control Lists.
+service WriteService {
+  option (google.api.default_host) = "keto.exampleapis.com";
+  // Writes one or more relation tuples in a single transaction.
+  rpc WriteRelationTuples(WriteRelationTuplesRequest) returns (WriteRelationTuplesResponse);
+}
+
+// The request of a WriteService.WriteRelationTuples rpc.
+message WriteRelationTuplesRequest {
+  // The write delta for the relation tuples operated in one single transaction.
+  // Either all actions commit or no change takes effect on error.
+  //
+  // The tuple's etag is used for a consistent read-modify-write request flow.
+  // If the etag is not present, the tuple is not validated against the state
+  // of the serverside tuple and get operated anyway.
+  repeated RelationTupleWriteDelta relation_tuple_deltas = 1;
+}
+
+// Write-delta for a WriteRelationTuplesRequest.
+message RelationTupleWriteDelta {
+  enum Action {
```

Comment:
Why would you add this action instead of having different rpcs? Is there a best practice regarding this defined somewhere?

Reply:
The are multiple reasons to this design:
- There would be quite a few and similar looking RPCs to cover each action.
- We can allow clients to do a transaction on multiple tuples in a single RPC call
  - This is very convenient for users in case the user wants all actions to succeed or none!
  - ...while also supporting behaviours like only insert this tuple if it does not already exists, otherwise I want the transaction to fail and give me an error
    - would be pretty easy to implement with SQL

Here are some similar refs:
- [IAM googleapis]( https://github.com/googleapis/googleapis/blob/3ac5ef0436d8dfeb2ca0091dc7fa8012da1c85af/google/iam/v1/policy.proto#L177)
- [GKE googleapis](https://github.com/googleapis/googleapis/blob/75d1324fff65abad6435e6d501e050439b5a616b/google/container/v1/cluster_service.proto#L2018)
- "Actions" can even be whole messages sometimes as in [Spanner googleapis](https://github.com/googleapis/googleapis/blob/b5653ef55579186841dee924592248d821902a12/google/spanner/v1/mutation.proto#L34)

We can always add RPC later if required.

Reply:
Or any other example where 2 or more modifications must succeed.

Imagine this (and I had this scenario learned the hard way):
- Some system inserts 3 tuples in response to an event (e.g. user registration...)
- The system continuously fails at the 3rd insert rpc.
- Now the system would need to rollback somehow and delete the first 2 inserted tuples for clean up.
- ...this is just a mess, cloud be solved client side, but only with much overhead, you get what I mean

- Otherwise, if a transaction is not needed, the user will just make multiple RPCs.

Reply:
:+1: agree, would really be awesome to have transactional requests here